File: local/src/lib/crdt/ccounter.js
// Optimized Add-Wins Observed-Remove Map

import { DotContext } from "../crdt/dotcontext";

export class CCounter {
    /**
     * The unique identifier for the current user
     * @type {string}
     */
    #id;

    /**
     * Map that stores [element_id, dot_value] => element
     */
    #map = new Map();

    /**
     * DotContext that stores element_id => dot_value
     */
    #cc = new DotContext();

    constructor(id) {
        this.#id = id;
    }

    /**
     *
     * @returns {Map} map
     */
    getMap() {
        const res = this.#map;
        return res;
    }

    /**
     *
     * @returns {DotContext} causal context
     */
    getCC() {
        const res = this.#cc;
        return res;
    }

    setID(id) {
        this.#id = id;
    }

    elements() {
        return this.#map.keys();
    }

    inc(value = 1) {
        let base = 0;
        for (const [key, _value] of this.#map) {
            if (key[0] == this.#id) { // Should only ever be one
                base = Math.max(base, _value);
                this.#map.delete(key);
            }
        }
        const dot = this.#cc.makeDot(this.#id);
        this.#map.set(dot, base + value);
    }

    dec(value = 1) {
        if (this.read() - value < 0) return;
        let base = 0;
        for (const [key, _value] of this.#map) {
            if (key[0] == this.#id) { // Should only ever be one
                base = Math.max(base, _value);
                this.#map.delete(key);
            }
        }
        const dot = this.#cc.makeDot(this.#id);
        this.#map.set(dot, base - value);
    }

    read() {
        let res = 0;
        for (const [key, value] of this.#map) {
            res += value;
        }
        return res;
    }

    /**
     * Checks if [key,value] is in the map
     * Note: I'm aware this is preposterous code but apparently two arrays ["banana", 1] and ["banana", 1] are not equal
     * 
     * @param {Array} key 
     * @returns 
     */
    hasKey(key) {
        for (const [_key, value] of this.#map) {
            if (_key[0] == key[0] && _key[1] == key[1]) return true;
        }
        return false;
    }

    /**
     * Join two CCounter
     *
     * @param {CCounter} ccounter
     */
    join(ccounter) {
        this.#map = new Map(
            [...this.#map].filter(
                ([key, value]) =>
                ccounter.hasKey(key) || !ccounter.getCC().dotIn(key)
            ),
        );

        for (const [key, value] of ccounter.getMap()) {
            if (!this.#cc.dotIn(key)) {
                this.#map.set(key, value);
            }
        }

        this.#cc.join(ccounter.getCC());
    }

    fromJSON(counter) {
        for (const count of counter.map) {
            const dot = [count.name, count.context];

            this.#map.set(dot, count.value);
        }

        // FIXME: this seems to be an edge case, where due to the fact that we are reconstructing a map from a JSON, we need to
        // insert the dots in the CC, but using insertDot() won't work due to how the compacting works, there may be a better way
        // to do this, or not it could be fine as it is as it doesn't seem to be a common issue
        for (const dot of counter.context.cc)
            this.#cc.getCC().set(dot[0], dot[1]);

        if (counter.context.dc.length > 0)
            for (const dot of counter.context.dc)
                this.#cc.insertDot(dot, false);
    }

    toJSON() {
        const res = {
            map: [],
            context: {},
        };

        for (const [key, value] of this.#map) {
            res.map.push({
                name: key[0],
                context: key[1],
                value: value,
            });
        }

        res.context = this.#cc.toJSON();

        return res;
    }
}

File: cloud/src/main/java/server/model/crdts/CCounter.java
package server.model.crdts;

import org.json.JSONArray;
import org.json.JSONObject;
import server.model.Product;
import server.model.utils.Pair;

import java.util.*;

public class CCounter {

    private String id; // unique user identifier
    private Map<Pair<String, Integer>, Integer> map = new HashMap<>();
    private DotContext cc = new DotContext();

    public CCounter(String id) {
        this.id = id;
    }

    public CCounter() {

    }

    public Map<Pair<String, Integer>, Integer> getMap() {
        return map;
    }

    public DotContext getCC() {
        return cc;
    }

    public void setID(String id) {
        this.id = id;
    }

    public Set<Pair<String, Integer>> elements() {
        return map.keySet();
    }

    public void inc(int value) {
        int base = 0;
        for (Map.Entry<Pair<String, Integer>, Integer> entry : map.entrySet()) {
            Pair<String, Integer> key = entry.getKey(); // [element_id, dot_value]
            int _value = entry.getValue(); // element
            if (key.getKey().equals(id)) { // Should only ever be one
                base = Math.max(base, _value);
                map.remove(key);
            }
        }
        Pair<String, Integer> dot = cc.makeDot(id);
        map.put(dot, base + value);
    }

    public void dec(int value) {
        if (this.read() - value < 0) return;

        int base = 0;
        for (Map.Entry<Pair<String, Integer>, Integer> entry : map.entrySet()) {
            Pair<String, Integer> key = entry.getKey(); // [element_id, dot_value]
            int _value = entry.getValue(); // element
            if (key.getKey().equals(id)) { // Should only ever be one
                base = Math.max(base, _value);
                map.remove(key);
            }
        }
        Pair<String, Integer> dot = cc.makeDot(id);
        map.put(dot, base - value);
    }

    public int read() {
        int res = 0;
        for (int value : map.values()) {
            res += value;
        }
        return res;
    }

    public boolean hasKey(Pair<String, Integer> key) {
        for (Map.Entry<Pair<String, Integer>, Integer> entry : map.entrySet()) {
            Pair<String, Integer> _key = entry.getKey();
            System.out.println(key + " == " + _key + "?");
            if (_key.equals(key)) {
                System.out.println("Yea!!");
                return true;
            }
            System.out.println("Nope!!");
        }
        return false;
    }

    public void join(CCounter ccounter) {
        for (Map.Entry<Pair<String, Integer>, Integer> entry : map.entrySet()) {
            Pair<String, Integer> key = entry.getKey();
            int value = entry.getValue();
            if (!ccounter.hasKey(key) && ccounter.getCC().dotIn(key)) {
                map.remove(key, value);
            }
        }

        for (Map.Entry<Pair<String, Integer>, Integer> entry : ccounter.getMap().entrySet()) {
            Pair<String, Integer> key = entry.getKey();
            Integer value = entry.getValue();
            if (!cc.dotIn(key)) {
                map.put(key, value);
            }
        }

        cc.join(ccounter.getCC());
    }

    public void fromJSON(JSONObject json) {

        for (Object product : json.getJSONArray("map")) {
            JSONObject jProduct = (JSONObject) product;
            Pair<String, Integer> dot = new Pair<>(jProduct.getString("name"), jProduct.getInt("context"));

            map.put(dot, (Integer) jProduct.get("value"));
        }

        for (Object dot : json.getJSONObject("context").getJSONArray("cc")) {
            JSONArray jDot = (JSONArray) dot;
            cc.getCC().put(jDot.getString(0), jDot.getInt(1)); // CHECKME
        }

        if (!json.getJSONObject("context").getJSONArray("dc").isEmpty()) {
            for (Object dot : json.getJSONObject("context").getJSONArray("dc")) {
                JSONArray jDot = (JSONArray) dot;
                Pair<String, Integer> pairDot = new Pair<>(jDot.getString(0), jDot.getInt(1));
                cc.insertDot(pairDot, false);
                // TESTME
            }
        }
    }

    public JSONObject toJSON() {
        JSONObject res = new JSONObject();
        ArrayList<JSONObject> map = new ArrayList<JSONObject>();

        for (Map.Entry<Pair<String, Integer>, Integer> entry : this.map.entrySet()) {
            JSONObject productJSON = new JSONObject();
            productJSON.put("name", entry.getKey().getKey());
            productJSON.put("context", entry.getKey().getValue());
            productJSON.put("value", entry.getValue());
            map.add(productJSON);
        }

        res.put("map", new JSONArray(map));
        res.put("context", this.cc.toJSON());

        return res;
    }
}
